File: rudolph-compiler/src/main/java/cn/wzbos/android/rudolph/AndroidModuleUtil.java
package cn.wzbos.android.rudolph;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;


/**
 * Create api module
 */
class AndroidModuleUtil {

    static void createApiModule(RudolphBuildInfo moduleInfo) {
        createGradle(moduleInfo);
        createAndroidManifest(moduleInfo);
    }

    /**
     * Create AndroidManifest.xml
     */
    private static void createAndroidManifest(RudolphBuildInfo info) {
        if (info != null && info.exportApi()) {
            File manifestFile = new File(info.projectPath + "/" + info.export_api_name + "/src/main/AndroidManifest.xml");
            if (!manifestFile.exists()) {
                if (!manifestFile.getParentFile().exists()) {
                    manifestFile.getParentFile().mkdirs();
                }
                FileWriter fileWriter = null;
                try {
                    if (manifestFile.createNewFile()) {
                        fileWriter = new FileWriter(manifestFile);
                        fileWriter.write("<manifest package=\"" + info.export_api_package + "\" />");
                        fileWriter.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    if (fileWriter != null) {
                        try {
                            fileWriter.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        }
    }

    /**
     * Create build.gradle
     */
    private static void createGradle(RudolphBuildInfo info) {
        File gradle_file = new File(info.projectPath + "/" + info.export_api_name + "/build.gradle");
        if (!gradle_file.exists()) {
            if (!gradle_file.getParentFile().exists()) {
                gradle_file.getParentFile().mkdirs();
            }
            FileWriter fileWriter = null;
            try {
                if (gradle_file.createNewFile()) {
                    fileWriter = new FileWriter(gradle_file);
                    fileWriter.write("/**\n" +
                            "* " + info.export_api_name + "\n" +
                            "* Generated code from Rudolph.\n" +
                            "* date : " + getCurrentDate() + "\n" +
                            "*/\n\n" +
                            "apply plugin: 'com.android.library'\n\n" +
                            "android {\n" +
                            "    compileSdkVersion " + info.compileSdkVersion + "\n\n" +
                            "    defaultConfig {\n" +
                            "        minSdkVersion " + info.minSdkVersion + "\n" +
                            "        targetSdkVersion " + info.targetSdkVersion + "\n" +
                            "        versionName " + info.versionName + "\n" +
                            "        versionCode " + info.versionCode + "\n" +
                            "    }\n" +
                            "}\n\n" +
                            "dependencies {\n" +
                            "    implementation fileTree(dir: 'libs', include: ['*.jar'])\n" +
                            "    implementation \"cn.wzbos.android:rudolph:$rudolph_version\"\n" +
                            "}");
                    fileWriter.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                if (fileWriter != null) {
                    try {
                        fileWriter.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }


    private static String getCurrentDate() {
        Date date = new Date(System.currentTimeMillis());
        return new SimpleDateFormat("yyyy/MM/dd HH:mm:ss").format(date);
    }
}

File: rudolph-compiler/src/main/java/cn/wzbos/android/rudolph/AndroidModuleUtil.kt
package cn.wzbos.android.rudolph

import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Create api module
 */
internal object AndroidModuleUtil {
    @JvmStatic
    fun createApiModule(moduleInfo: RudolphBuildInfo) {
        createGradle(moduleInfo)
        createAndroidManifest(moduleInfo)
    }

    /**
     * Create AndroidManifest.xml
     */
    private fun createAndroidManifest(info: RudolphBuildInfo?) {
        if (info != null && info.exportApi()) {
            val manifestFile = File(info.projectPath + "/" + info.exportProtocolName + "/src/main/AndroidManifest.xml")
            if (!manifestFile.exists()) {
                if (!manifestFile.parentFile.exists()) {
                    manifestFile.parentFile.mkdirs()
                }
                var fileWriter: FileWriter? = null
                try {
                    if (manifestFile.createNewFile()) {
                        fileWriter = FileWriter(manifestFile)
                        fileWriter.write("<manifest package=\"" + info.exportProtocolPackage + "\" />")
                        fileWriter.close()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                } finally {
                    if (fileWriter != null) {
                        try {
                            fileWriter.close()
                        } catch (e: IOException) {
                            e.printStackTrace()
                        }
                    }
                }
            }
        }
    }

    /**
     * Create build.gradle
     */
    private fun createGradle(info: RudolphBuildInfo) {
        val gradleFile = File(info.projectPath + "/" + info.exportProtocolName + "/build.gradle")
        if (!gradleFile.exists()) {
            if (!gradleFile.parentFile.exists()) {
                gradleFile.parentFile.mkdirs()
            }
            var fileWriter: FileWriter? = null
            try {
                if (gradleFile.createNewFile()) {
                    fileWriter = FileWriter(gradleFile)
                    fileWriter.write("""/**
* ${info.exportProtocolName}
* Generated code from Rudolph.
* date : $currentDate
*/

apply plugin: 'com.android.library'

android {
    compileSdkVersion ${info.compileSdkVersion}

    defaultConfig {
        minSdkVersion ${info.minSdkVersion}
        targetSdkVersion ${info.targetSdkVersion}
        versionName ${info.versionName}
        versionCode ${info.versionCode}
    }
}

dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation "cn.wzbos.android:rudolph:${"$"}rudolph_version"
}""")
                    fileWriter.close()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                if (fileWriter != null) {
                    try {
                        fileWriter.close()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }
        }
    }

    private val currentDate: String
        get() {
            val date = Date(System.currentTimeMillis())
            return SimpleDateFormat("yyyy/MM/dd HH:mm:ss").format(date)
        }
}